### PR TITLE
Migrate Source Filter Screen to Compose

### PR DESCRIPTION
**Pull request by FourTOne5**

### CODE REVIEW

File: app/src/main/java/eu/kanade/domain/source/interactor/GetSourceFilterData.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                        FilterUiModel.Item(
+                            source,
+                            source.id.toString() !in disabledSource
+                        )
+                    }.also { models ->
+                        items.addAll(models)
+                    }
+                }
+                items
+            }
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/domain/source/interactor/GetSourceFilterData.kt:51]:
```suggestion
                val isLangEnabled = it.key in enabledLanguage
                val items = listOf(FilterUiModel.Header(it.key, isLangEnabled))
                if (isLangEnabled.not()) return@flatMap items
                items + it.value.map { source ->
                    FilterUiModel.Item(
                        source,
                        source.id.toString() !in disabledSource
                    )
                }
```
I would also move the flatmap to presentation layer

File: app/src/main/java/eu/kanade/tachiyomi/source/SourceManager.kt [hunk 2]
```diff
@@ -18,6 +19,8 @@ open class SourceManager(private val context: Context) {
 
     private val _catalogueSources: MutableStateFlow<List<CatalogueSource>> = MutableStateFlow(listOf())
     val catalogueSources: Flow<List<CatalogueSource>> = _catalogueSources
+    val onlineSources: Flow<List<HttpSource>> =
+        _catalogueSources.map { sources -> sources.filterIsInstance<HttpSource>() }
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/tachiyomi/source/SourceManager.kt:23]:
```suggestion
    val onlineSources: Flow<List<HttpSource>> = _catalogueSources.filterIsInstance()
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterController.kt [hunk 3]
```diff
@@ -1,112 +1,52 @@
...
+    override fun ComposeContent(nestedScrollInterop: NestedScrollConnection) {
+        SourceFilterScreen(
+            nestedScrollInterop = nestedScrollInterop,
+            presenter = presenter,
+            onClickLang = { lang, isChecked ->
+                if (isChecked) {
+                    preferences.enabledLanguages() += lang
+                } else {
+                    preferences.enabledLanguages() -= lang
                 }
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterController.kt:36]: This should be in the presenter and `isChecked` isn't really needed as we can check what `enabledLanguages` contains. Could also be an interactor as extension language filter will also Toggle langauges

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterController.kt [hunk 4]
```diff
@@ -1,112 +1,52 @@
...
-                        addLanguageSources(this, sources)
-                    }
-                    true
+            },
+            onClickSource = { source, isSourceEnabled ->
+                if (isSourceEnabled) {
+                    preferences.disabledSources() -= source.id.toString()
+                } else {
+                    preferences.disabledSources() += source.id.toString()
                 }
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterController.kt:43]: This should be in the presenter and `isSourceEnabled` isn't really needed as we can check what `disabledSources` contains. Could probably change `DisableSource` to be a toggle `ToggleSource` that way we can reuse the logic in both this screen and for the Source screen

File: app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        modifier = modifier
+    ) {
+        PreferenceRow(
+            title = LocaleHelper.getSourceDisplayName(language, LocalContext.current),
+            action = {
+                Switch(checked = isEnabled, onCheckedChange = null)
+            },
+            onClick = { onClickItem(language, !isEnabled) },
+        )
+    }
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt:114]: @arkon maybe `PreferenceRow` should have a way to modify the modifier?

File: app/src/main/java/eu/kanade/tachiyomi/source/SourceManager.kt (hunk 2 above)
Comment (FourTOne5) [reply app/src/main/java/eu/kanade/tachiyomi/source/SourceManager.kt:23]: I tried it but it didn't work Local Source was also appearing there.

Comment (ghostbear) [reply app/src/main/java/eu/kanade/tachiyomi/source/SourceManager.kt:23]: Alright keep it as is then

File: app/src/main/java/eu/kanade/domain/source/interactor/GetSortedSourceLangMap.kt [hunk 6]
```diff
@@ -3,17 +3,16 @@ package eu.kanade.domain.source.interactor
 import eu.kanade.domain.source.model.Source
 import eu.kanade.domain.source.repository.SourceRepository
 import eu.kanade.tachiyomi.data.preference.PreferencesHelper
-import eu.kanade.tachiyomi.ui.browse.source.FilterUiModel
 import eu.kanade.tachiyomi.util.system.LocaleHelper
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.combine
 
-class GetSourceFilterData(
+class GetSortedSourceLangMap(
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/domain/source/interactor/GetSortedSourceLangMap.kt:10]: `GetSortedSourceLangMap` -> `GetLangaugesWithSources`?

Comment (FourTOne5) [reply app/src/main/java/eu/kanade/domain/source/interactor/GetSortedSourceLangMap.kt:10]: That sounds better

File: app/src/main/java/eu/kanade/domain/source/interactor/ToggleLanguage.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import eu.kanade.tachiyomi.util.preference.minusAssign
+import eu.kanade.tachiyomi.util.preference.plusAssign
+
+class ToggleLanguage(
+    val preferences: PreferencesHelper
+) {
+
+    fun await(language: String) {
+        val isLangEnabled = language in preferences.enabledLanguages().get()
+        if (isLangEnabled) {
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/domain/source/interactor/ToggleLanguage.kt:13]:
```suggestion
        val isEnabled = language in preferences.enabledLanguages().get()
        if (isEnabled) {
```

File: app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                    isEnabled = model.isEnabled,
+                    onClickItem = onClickSource
+                )
+            }
+        }
+    }
+}
+
+@Composable
+fun LangHeader(
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt:98]:
```suggestion
fun SourceFilterHeader(
```

File: app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            action = {
+                Switch(checked = isEnabled, onCheckedChange = null)
+            },
+            onClick = { onClickItem(language) },
+        )
+    }
+}
+
+@Composable
+fun LangItem(
```
Comment (ghostbear) [inline app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt:118]:
```suggestion
fun SourceFilterItem(
```

**Review by ghostbear — APPROVED** [review]

File: app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt (hunk 5 above)
Comment (arkon) [reply app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt]: Now you can.

**Comment by FourTOne5** [comment]
Gonna fix conflict in a bit

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    fun toggleSource(source: Source) {
+        toggleSource.await(source)
+    }
+
+    fun toggleLanguage(language: String) {
+        toggleLanguage.await(language)
+    }
+}
+
+data class SourceFilterState(
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterPresenter.kt:72]: Use a sealed class similar to `MigrateMangaState`

File: app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+import eu.kanade.tachiyomi.ui.browse.source.FilterUiModel
+import eu.kanade.tachiyomi.ui.browse.source.SourceFilterPresenter
+import eu.kanade.tachiyomi.util.system.LocaleHelper
+
+@Composable
+fun SourceFilterScreen(
+    nestedScrollInterop: NestedScrollConnection,
+    presenter: SourceFilterPresenter,
+    onClickLang: ((String) -> Unit),
+    onClickSource: ((Source) -> Unit)
```
Comment (arkon) [inline app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt:34]: Don't think you need the extra brackets

File: app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+) {
+    val state by presenter.state.collectAsState()
+    val items = state.uiModels.orEmpty()
+
+    when {
+        state.isLoading -> LoadingScreen()
+        state.hasError -> Text(text = state.error!!.message!!)
+        state.isEmpty -> EmptyScreen(message = "")
```
Comment (arkon) [inline app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt:42]: Should have some sort of message?

File: app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        )
+    }
+}
+
+@Composable
+fun SourceFilterContent(
+    nestedScrollInterop: NestedScrollConnection,
+    items: List<FilterUiModel>,
+    onClickLang: ((String) -> Unit),
+    onClickSource: ((Source) -> Unit)
```
Comment (arkon) [inline app/src/main/java/eu/kanade/presentation/source/SourceFilterScreen.kt:57]: Unnecessary brackets

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterPresenter.kt [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+import uy.kohesive.injekt.api.get
+import uy.kohesive.injekt.injectLazy
+
+class SourceFilterPresenter(
+    private val getLanguagesWithSources: GetLanguagesWithSources = Injekt.get(),
+    private val toggleSource: ToggleSource = Injekt.get(),
+    private val toggleLanguage: ToggleLanguage = Injekt.get()
+) : BasePresenter<SourceFilterController>() {
+
+    val preferences: PreferencesHelper by injectLazy()
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterPresenter.kt:27]: Can inject this in the constructor as well.

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterPresenter.kt [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    private val _state: MutableStateFlow<SourceFilterState> = MutableStateFlow(SourceFilterState.LOADING)
+    val state: StateFlow<SourceFilterState> = _state.asStateFlow()
+
+    override fun onCreate(savedState: Bundle?) {
+        super.onCreate(savedState)
+        presenterScope.launchIO {
+            getLanguagesWithSources.subscribe()
+                .catch { exception ->
+                    _state.update { state ->
+                        state.copy(uiModels = listOf(), error = exception)
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceFilterPresenter.kt:38]: Use `emptyList()`

**Pull request merged**